source/extensions/io_socket/user_space/io_socket_handle_impl.cc

```diff
@@ -0,0 +1,330 @@
+                                                    const Network::Address::Instance&) {
+  return Network::IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result IoSocketHandleImpl::recvmsg(Buffer::RawSlice*, const uint64_t, uint32_t,
+                                                    RecvMsgOutput&) {
+  return Network::IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result IoSocketHandleImpl::recvmmsg(RawSliceArrays&, uint32_t, RecvMsgOutput&) {
+  return Network::IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result IoSocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  if (!isOpen()) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // No data and the writer closed.
+  if (pending_received_data_.length() == 0) {
+    if (receive_data_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                                 Network::IoSocketError::deleteIoError)};
+    }
+  }
+  // Specify uint64_t since the latter length may not have the same type.
+  auto max_bytes_to_read = std::min<uint64_t>(pending_received_data_.length(), length);
+  pending_received_data_.copyOut(0, max_bytes_to_read, buffer);
+  if (!(flags & MSG_PEEK)) {
+    pending_received_data_.drain(max_bytes_to_read);
+  }
+  return {max_bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+bool IoSocketHandleImpl::supportsMmsg() const { return false; }
+
+bool IoSocketHandleImpl::supportsUdpGro() const { return false; }
+
+Api::SysCallIntResult IoSocketHandleImpl::bind(Network::Address::InstanceConstSharedPtr) {
+  return makeInvalidSyscallResult();
+}
+
+Api::SysCallIntResult IoSocketHandleImpl::listen(int) { return makeInvalidSyscallResult(); }
+
+Network::IoHandlePtr IoSocketHandleImpl::accept(struct sockaddr*, socklen_t*) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+Api::SysCallIntResult IoSocketHandleImpl::connect(Network::Address::InstanceConstSharedPtr) {
+  // Buffered Io handle should always be considered as connected.
+  // Use write or read to determine if peer is closed.
+  return {0, 0};
+}
+
+Api::SysCallIntResult IoSocketHandleImpl::setOption(int, int, const void*, socklen_t) {
+  return makeInvalidSyscallResult();
+}
+
+Api::SysCallIntResult IoSocketHandleImpl::getOption(int, int, void*, socklen_t*) {
+  return makeInvalidSyscallResult();
+}
+
+Api::SysCallIntResult IoSocketHandleImpl::setBlocking(bool) { return makeInvalidSyscallResult(); }
+
+absl::optional<int> IoSocketHandleImpl::domain() { return absl::nullopt; }
+
+Network::Address::InstanceConstSharedPtr IoSocketHandleImpl::localAddress() {
+  // TODO(lambdai): Rewrite when caller accept error as the return value.
```

Comment:
@asraa @yanavlasov 

I'm not parsing this TODO.  We should file an issue about exception use in localAddress()

Reply:
By going through how Envoy will use, I find it appropriate to return a dummy local address at IoHandleImpl

Envoy then should use Socket to wrap this IoHandle. In That Socket an addressProvider should be provided. The listener filter/network filter should use that addressProvider instead of the low level `IoHandleImpl`

Reply:
Returning valid EnvoyInternalAddress in commit https://github.com/envoyproxy/envoy/pull/14917/commits/8e5a5a59586d28c1a5fecf62c1b388fa67ba7462